### PR TITLE
Move futures out of kernel

### PR DESCRIPTION
**Pull request by ppedrot**
This PR removes the use of the delayed future mechanism from the kernel.

The table of opaque proofs now lives in vernac/opaques and we fully inverted control in the kernel. The new kernel API is push-based, i.e. in order to declare a delayed proof in the kernel one must first register an entry, returning a handle that can later be used to fill in the missing proof. Pushing an opaque proof inside the kernel is itself a two-step process. First, one must provide an actual proof term that will be type-checked. This returns an abstract certificate that can be used in a second step to mark the proof as valid without having to replay the type-checking process.

By making the API purely functional some oddities have been made explicit. The main one is that joining an environment in the STM is actually a no-op w.r.t. the global state. It only checks that the current set of delayed constraints is consistent, but it does not push them into the global environment because joining is not part of the VCS. This was hidden by the laziness of the Future mechanism before the PR and ought to be handled in a more principled way. In practice that means that in interactive mode the UI can display all Qed proofs as processed but their global universe constraints are actually not part of the current environment. The only point in time where this hold is right after joining the environment. Any subsequent action may desynchronize the constraints. This should probably be tweaked in a later patch, but it requires taming the STM first.

Further cleanups are now possible. Since futures are now right next to the STM we can hope for a simpler implementation.

Overlays:
- https://github.com/coq-community/paramcoq/pull/80
- https://github.com/MetaCoq/metacoq/pull/588

### CODE REVIEW

**Review by ejgallego — APPROVED** [review]
Impressive! For now seems more than good [modulo bugs]

File: checker/check.ml
Comment (Alizter) [reply checker/check.ml:108]: But maybe I am mistaken, do we not end errors with full stops?

File: kernel/opaqueproof.mli
Comment (ppedrot) [reply kernel/opaqueproof.mli:30]: Nope, will remove it.

File: kernel/safe_typing.ml
Comment (ppedrot) [reply kernel/safe_typing.ml:141]: I actually didn't know the slang meaning, but not all dictionaries I've looked at actually mention it. Meanwhile it's a very standard crypto concept, so I don't think this will be causing trouble.

**Comment by SkySkimmer** [comment]
@coqbot ci minimize bedrock2

**Comment by ppedrot** [comment]
@SkySkimmer I can't reproduce the bedrock2 failure locally. I think it's just an OOM that was barely passing before, but now consumes slightly more memory and thus dies.

**Comment by SkySkimmer** [comment]
Are you benching this btw!?

File: checker/check.ml
Comment (ppedrot) [reply checker/check.ml:108]: I don't know, it's not consistent across the codebase.

**Comment by ppedrot** [comment]
Bench: https://gitlab.com/coq/coq/-/jobs/1620249506

**Comment by ppedrot** [comment]
Note that looking a bit at bedrock2 showed me that Print Assumptions is completely crazy, since it does not cache the contents of the body of a global. This is not specific to this PR, it can already be very slow on master since that means we keep fetching and substituting huge proof terms for nothing. Maybe this patch is making it marginally worse but there is a clear and small unrelated fix that I'll submit separately.

**Comment by SkySkimmer** [comment]
We don't bench bedrock2 anyway (the bench claims to do it, but it's actually a tiny subset as you can tell by it taking less than 2min)

**Comment by ppedrot** [comment]
Wait a sec, I think I know why bedrock2 explodes midflight. I've just realized that I forgot to perform hashconsing on opaque proofs in the upper layers. That probably explains why the bench is faster. Out of curiosity I'll still let it run to completion before fixing the PR.

**Review by gares — COMMENTED** [review]
LGTM modulo the question on the Global.safe_env call in the thunk

File: vernac/opaques.ml
Comment (ppedrot) [reply vernac/opaques.ml:75]: > Can you explain why you want Global.safe_env to be called when the thunk is forced

I actually wondered about that when writing it. It shouldn't change anything, because the set of futures of the environment at that time will be an extension of the one at definition time. In the current phrasing, we get an anomaly if somebody tries to check an opaque after it has been filled, which is an additional safety. That's why I preferred it that way.

> Also defined_opaque seems like an oxymoron... maybe use proved instead?

Unfortunately, "defined" is quite polysemic. Not sure "proved" is any better, but I do need a name to make the distinction with the private case...

Comment (gares) [reply vernac/opaques.ml:75]: The problem is that defined recalls `Defined.` which is not opaque.

About the safe_env, if it changes anything, then passing it explicitly would avoid any question.
My problem with the current code is that you need to understand how the table of thunks is synchronized w.r.t. Global in order to see that "it changes nothing". IMO it would be a better API to explicitly say in which env one wants the promise to be checked.

**Comment by gares** [comment]
General question: in the old API the safe_env opaque type had a method to ensure that all delayed proofs were there and checked.
This API was part of the TCB.

I can't find the corresponding entry in the TCB. Can you point it out?

**Comment by ppedrot** [comment]
> I can't find the corresponding entry in the TCB. Can you point it out?

It didn't move, it's still `Safe_env.is_joined_environment`.

**Comment by gares** [comment]
> It didn't move, it's still `Safe_env.is_joined_environment`.

Great, but that one is not called (in master) whenever we save a .vo library to disk (while the one you moved to upper layers yes). IMO this one should be called, that is the tool that saves a .vo should ask the kernel if the env is joined.

**Comment by ppedrot** [comment]
I've added a check that the env is joined in the commit about assertions.

File: kernel/safe_typing.ml
Comment (ppedrot) [reply kernel/safe_typing.ml:1032]: I didn't write this piece of code, and I hate this style that does not scale with a passion. The problem is that fixing it requires touching unrelated lines of code and produces an ugly diff.

File: vernac/opaques.ml
Comment (ppedrot) [reply vernac/opaques.ml:75]: > My problem with the current code is that you need to understand how the table of thunks is synchronized

Actually, it's much worse than that. You need to understand the STM at all to know the invariants and it takes a naughty pleasure to silently remove the rug under your feet. If anything I'd rather have a purely functional interface but unfortunately this requires changes that are way beyond this PR. There are very weird things going on but we're now entering the STM twilight zone.

**Comment by ejgallego** [comment]
> Bench: https://gitlab.com/coq/coq/-/jobs/1620249506

Given this, having a flag `-i-dont-care-about-memory-but-i-do-care-about speed` could be very interesting for some users.

**Comment by ppedrot** [comment]
Bench with hashconsing: https://gitlab.com/coq/coq/-/jobs/1621402930

**Comment by gares** [comment]
@ppedrot I don't get why odd-order gets 16% more memory use without hashconsing. I thought we did hashcons only opaque terms (and for sure there is no extraction or Print which loads them).

**Comment by ejgallego** [comment]
> @ppedrot I don't get why odd-order gets 16% more memory use without hashconsing. I thought we did hashcons only opaque terms (and for sure there is no extraction or Print which loads them).

Could indeed this be due to actually 16% sharing inside opaque proofs?

**Comment by gares** [comment]
That could be (just holding them until the file ends). If so, the option "-dont-store-opaque-proofs" would be even better.

**Comment by ppedrot** [comment]
> I don't get why odd-order gets 16% more memory use without hashconsing.

@gares I am not sure what seems surprising to you there. Were you expecting less or more memory use without hashconsing? Obviously since we don't hashcons proof terms take more space in memory, leading to the observed increase...

**Comment by gares** [comment]
In general I'm puzzled the overhead is so little, so why bother?!
I'm puzzled to see it in the OOthm, maybe there proofs are super large but then.. only 15% is saved by maximal sharing?

Moreover unimath gets a 50% extra space, but IIRC it has no opaque proofs. Do we share transparent terms?

It is OT for this PR, but these numbers don't make much sense to me. It seems a feature one may opt in, but not a sane default.

**Comment by SkySkimmer** [comment]
>In general I'm puzzled the overhead is so little, so why bother?!

There is also vo size

>Moreover unimath gets a 50% extra space, but IIRC it has no opaque proofs. Do we share transparent terms?

https://github.com/UniMath/UniMath/search?q=qed

**Comment by ppedrot** [comment]
We do share transparent proofs as well.

**Comment by ppedrot** [comment]
Also note that the memory measure of the process is not reliable, it's common to see it vary up to 30% without any good reason. We should use instead the informations coming directly from the GC as per the `Print Debug GC` command.

File: vernac/opaques.ml
Comment (gares) [reply vernac/opaques.ml:75]: OK, so pass the safe env explicitly. In that case "which env is used to typecheck" does not depend on when the thunk is forced.

Comment (ppedrot) [reply vernac/opaques.ml:75]: As argued above, w.r.t. to invariants I believe that the current situation is better. Furthermore passing the env explicitly would keep a useless pointer to an old environment for no good reason. Also, it makes the interface looks like it's purely functional, but it's not...

**Review by gares — APPROVED** [review]
ok then, but please leave a comment about the save env and the thunk firing time

File: kernel/safe_typing.ml
Comment (ppedrot) [reply kernel/safe_typing.ml:1040]: Weird, it does look like I introduced this code, you're right. Maybe a leftover of a rebase? In any case I'll remove that.

Comment (ppedrot) [reply kernel/safe_typing.ml:1040]: (Definitely a rebase issue over 7a4e7df).

**Comment by ppedrot** [comment]
@gares comments addressed.

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
I will look into the async failure.
I need a review from @coq/kernel-maintainers.
I'm running a bench https://gitlab.com/coq/coq/-/jobs/1627376688

**Comment by ppedrot** [comment]
The async failure is unrelated, it's been introduced by #14758. Also, we already had a bench there: https://github.com/coq/coq/pull/14924#issuecomment-926872725, there was no code change in between.

**Comment by gares** [comment]
> The async failure is unrelated, it's been introduced by #14758.

You mean that the PR adds a test with *the* case where processing proof in async mode (adding late univ constraints from opaque body) does matter for TC search? If so, @SkySkimmer you should add `-async-proofs off` to `bug_14734.v` (or even better implement `assert_fails_univ_async`).

**Comment by ppedrot** [comment]
> You mean that the PR adds a test with the case where processing proof in async mode (adding late univ constraints from opaque body) does matter for TC search?

Erm, actually yes.

> or even better implement assert_fails_univ_async

Couldn't we also have a few vernacular commands that act as write barriers and attributes to control finely the deferring of Qed-opaque proofs? For instance I started implementing a mechanism for private monomorphic universes and for efficient document processing it'd be nice to flag Qed proofs that are known to introduce new constraints between global universes.

File: stm/stm.ml
Comment (SkySkimmer) [inline stm/stm.ml:1686]: Why not remove this now useless match?

File: kernel/safe_typing.ml
Comment (SkySkimmer) [inline kernel/safe_typing.ml:935]: When do we get Not_found here? Should we anomaly instead of returning the input env?

File: kernel/safe_typing.ml
Comment (SkySkimmer) [inline kernel/safe_typing.ml:1145]: Why don't we just 
```suggestion
    future_cst = senv.future_cst;
```
here?

If that's incorrect why do we fold add instead of union?

**Review by SkySkimmer — APPROVED** [review]

File: kernel/safe_typing.ml
Comment (ppedrot) [reply kernel/safe_typing.ml:935]: No, this code path is exercised. The problem is that the join operation is not in the CVS tree, so if you do it twice you'll get an anomaly. Joining once is essentially introducing a desynchronization between the safe env and the opaque table, but it does not matter because the only discrepancy is that the opaque table forgets that the safe env has been filled. To fix this we should mark somehow that a join was performed somewhere in the structure so that it is remember correctly.

File: stm/stm.ml
Comment (ppedrot) [reply stm/stm.ml:1686]: It's still checking the assertion though.

Comment (SkySkimmer) [reply stm/stm.ml:1686]: Is the assert a check or just because we have to put something in that branch?
If it's a check then change the match to return unit.

Comment (ppedrot) [reply stm/stm.ml:1686]: I don't know what was the original intent, but probably both. I'll turn it into a unit value then.

File: kernel/safe_typing.ml
Comment (ppedrot) [reply kernel/safe_typing.ml:1145]: To be fair I just adapted the code without much thought. I have no idea why we have not two, but *three* different safe environments in this function and what invariants relate them. As for `union` it's just that fold + add seems more natural to me that the weird union API for maps, but this is essentially a mater of taste and there is no serious reason otherwise.

File: kernel/safe_typing.ml
Comment (ppedrot) [reply kernel/safe_typing.ml:935]: The comment is precisely about that, since I don't know of the best API to perform the double check. The various ways I see are: returning an option / providing a `is_filled` primitive / raising a particular exception. All of these choices have some drawbacks and I can't decide for any. Also there is only one caller of `fill_opaque` so it's somehow introducing quite a tight coupling. I'll try to write a small enhancement anyways.

File: kernel/safe_typing.ml
Comment (gares) [reply kernel/safe_typing.ml:1145]: This function comes from here
https://github.com/coq/coq/commit/fac949450909b5ef17078f220ae809cf54ae3f08#diff-24b536f3006f7ced4984e883e87dbe8a0a1b0b6c921bf661f7f307c41bd88d53R558

File: checker/check.ml
Comment (Zimmi48) [reply checker/check.ml:108]: Well, the answer is yes, and we are trying to be consistent.

File: kernel/safe_typing.ml
Comment (ppedrot) [reply kernel/safe_typing.ml:1145]: If I read the code correctly, this function is actually buggy in master. Namely, joining the environment inside a module is not really joining it since the delayed constraints from outside the module being currently defined are not added to the global environment. This bug transposes in this PR with an anomaly when joining from withing a module. Small reproducible example:
```
Lemma foo : True.
Proof.
constructor.
Qed.

Module Foo.
(* Joining here results in 'Anomaly "Missing opaque with identifier 0"' *)
```
I am not sure that claiming we're correct while silently ignoring universe constraints is a good behaviour either, but at least it's less visible to the end user. I'll fix this with an additional patch.

Comment (gares) [reply kernel/safe_typing.ml:1145]: Well, in some sense it is not a bug, since you can't save a vo file with an open module. But I prefer the old behavior (no failure, no guarantee) to the new anomaly. Can't you turn it into a better error?

Comment (ppedrot) [reply kernel/safe_typing.ml:1145]: > Well, in some sense it is not a bug

This further weakens the expected invariant of joining the environment. I thought it was "things make sense just right after you joined the env", but it seems you also need to add a bunch of provisos like "no modules are opened" and "the age of the captain is 42". I don't mind trying to be more lenient, but at the very least we need an actual spec.

Comment (SkySkimmer) [reply kernel/safe_typing.ml:1145]: Can't we just get joining to work in modules? Is there more to do than add `future_cst = senv.future_cst;` in start_module/modtype?

Comment (ppedrot) [reply kernel/safe_typing.ml:1145]: I don't think so, that's what I've been starting writing.

Comment (gares) [reply kernel/safe_typing.ml:1145]: >  "no modules are opened" and "the age of the captain is 42".

I guess you could also make it work with an open module, but it is more work for very little in return IMO.
Turning the anomaly into "sorry, you can't join the env with an open module" is much simpler and user friendly.

Comment (ppedrot) [reply kernel/safe_typing.ml]: It's trivial now, I've just added a patch for that. That said I do not manage to get a working `.fake` test for that situation, see the last commit (0355b359594f3be02b67e5c0bafd13696acfafe4). It fails with
```
...
FAIL 0 [Universe inconsistency. Cannot enforce v < u because u < v.]
fake_ide: error: Universe inconsistency. Cannot enforce v < u because u < v.
The command failed as expected
fake_ide: error: No state found
```
That's produced by `FAILJOIN` so it recognizes that a failure is expected but then it behaves weirdly w.r.t. the expected state.

Comment (gares) [reply kernel/safe_typing.ml]: Hum, IMO the new join does not get attached (by the STM?) a last_good_id, so it backtracks to the initial state or something like that. I can have a look. What happens in CoqIDE when you hit the wheels button?

Comment (ppedrot) [reply kernel/safe_typing.ml]: It displays an error in the right pane, but does not backtrack anywhere. Any subsequent join fails as well.

**Comment by ppedrot** [comment]
@gares I've hardened the opaque filling API as you requested.

File: kernel/safe_typing.ml
Comment (gares) [reply kernel/safe_typing.ml]:
```
gares@ollypat:~/COQ/move-opaque-out-of-kernel$ test-suite/ide/fake_ide.exe _build/install/default/bin/coqidetop.opt test-suite/ide/join-module2.fake 
Running: _build/install/default/bin/coqidetop.opt --xml_format=Ppcmds
Add ((" Universes u v. ",-1),(1,true))
GOOD (2,(Inl ,""))
Add ((" Lemma foo : True. ",-2),(2,true))
GOOD (3,(Inl ,""))
Add ((" Proof. ",-3),(3,true))
GOOD (4,(Inl ,""))
Add ((" refine (let _ : Type@{v} := Type@{u} in _). ",-4),(4,true))
GOOD (5,(Inl ,""))
Add ((" constructor. ",-5),(5,true))
GOOD (6,(Inl ,""))
Add ((" Qed. ",-6),(6,true))
GOOD (7,(Inl ,""))
Add ((" Lemma bar : True. ",-7),(7,true))
GOOD (8,(Inl ,""))
Add ((" Proof. ",-8),(8,true))
GOOD (9,(Inl ,""))
Add ((" refine (let _ : Type@{u} := Type@{v} in _). ",-9),(9,true))
GOOD (10,(Inl ,""))
Add ((" constructor. ",-10),(10,true))
GOOD (11,(Inl ,""))
Add ((" Qed. ",-11),(11,true))
GOOD (12,(Inl ,""))
Add ((" Module Foo. ",-12),(12,true))
GOOD (13,(Inl ,""))
Status true
FAIL 9 (1,43)[The term "Type" has type "Type@{v+1}" while it is expected to have type
 "Type@{u}" (universe inconsistency: Cannot enforce v < u because u < v).]
fake_ide: error: The term "Type" has type "Type@{v+1}"
                 while it is expected to have type "Type@{u}"
                 (universe inconsistency: Cannot enforce v < u because u
                 < v).
The command failed as expected
Edit_at 9
GOOD Inl 

*[          ,   2] Universes u v. 
*[          ,   3] Lemma foo : True. 
*[          ,   4] Proof. 
*[          ,   5] refine (let _ : ...
*[          ,   6] constructor. 
*[          ,   7] Qed. 
*[          ,   8] Lemma bar : True. 
*[          ,   9] Proof. 
Quitting fake_ide
gares@ollypat:~/COQ/move-opaque-out-of-kernel$ echo $?
0
```
How did you get the error?

Comment (gares) [reply kernel/safe_typing.ml]: As you can see, I get a `Fail 9`, which means go back to state 9, the last known good one. And it seems to go there. CoqIDE should backtrack there as well.

Maybe you did fix the bug without realizing it?

Comment (ppedrot) [reply kernel/safe_typing.ml]: Running your command works with the same output as yours, but running the test-suite still fails on this example. What's going on?

Comment (ppedrot) [reply kernel/safe_typing.ml]: Running your command works with the same output as yours, but running the test-suite still fails on this example. What's going on?

Comment (ppedrot) [reply kernel/safe_typing.ml]: @gares any update on this?

Comment (gares) [reply kernel/safe_typing.ml]: No sorry, I had no time to look into this. I'll do it now.

Comment (gares) [reply kernel/safe_typing.ml]: So
```
(cd test-suite/; ide/fake_ide.exe '/home/gares/COQ/move-opaque-out-of-kernel'/_build/install/default/bin/coqidetop.opt "ide/join-module2.fake" "-q -async-proofs on -async-proofs-tactic-error-resilience off -async-proofs-command-error-resilience off ")
```
fails! Apparently `-async-proofs off` is the default for fake_ide (and not for CoqIDE). Which is silly² since the whole point of fake_ide is to test CoqIDE as it is... so the bug is real.

Comment (ppedrot) [reply kernel/safe_typing.ml]: > so the bug is real.

That's a net progress, but this does not explain the bug, unfortunately...

Comment (gares) [reply kernel/safe_typing.ml]:
```
Status true
FAIL 0 [Universe inconsistency. Cannot enforce v < u because u < v.]
fake_ide: error: Universe inconsistency. Cannot enforce v < u because u < v.
The command failed as expected
fake_ide: error: No state found
```
As I was saying, this means that the error does not carry a `safe_id`, 0 is the dummy id. So jumping back to 0 is an error. The STM was passing a `fix_exn` argument to `Future.*` in order to enrich any exception escaping the thunk with this piece of info. I don't know how this is handled now, since I have some memory of this argument going away at some point. I'll check it, but if this ring a bell please tell me.

File: kernel/safe_typing.ml
Comment (SkySkimmer) [inline kernel/safe_typing.ml:919]: strange indentation btw

File: kernel/safe_typing.ml
Comment (gares) [reply kernel/safe_typing.ml]: Here:
```
        let cert0 = Future.join cert in
        let () = Global.fill_opaque cert0 in
```
You force the thunk, which hopefully would raise exceptions enriched the right way. But the you check outside the thing that `cert0` is ok, so IMO this check does not raise a correctly wrapped exception. (Maybe the bug was already there). Does this sound like a plausible explanation?

Comment (ppedrot) [reply kernel/safe_typing.ml]: I see, so that probably means I have to wrap the exception coming from the join operation with the function provided with the corresponding future in vernac/opaques. I'll try that.

Comment (gares) [reply kernel/safe_typing.ml]: If I'm right, `Future.chain cert (fun cert0 -> Global.fill_opaque cert0)` should do it...

Comment (ppedrot) [reply kernel/safe_typing.ml]: Indeed, something like that. It's a bit ugly to have to work around the Future API but for the time being that'll do it.

Comment (ppedrot) [reply kernel/safe_typing.ml]: It seems to have worked, miraculously...

File: kernel/safe_typing.ml
Comment (ppedrot) [reply kernel/safe_typing.ml]: Fixed.

**Comment by ppedrot** [comment]
@gares Thanks a lot, it seems this is fine now. I think that the STM API could be simplified quite a bit now that the futures are only in the upper layers, allowing for finer-grained invariants (and lesser reliance on hacks like `fix_exn`, no offense). But for now I think this PR can be considered ready.

**Comment by gares** [comment]
Great! But I don't see how to make the fix_exn hack less hackish, so suggestions are welcome. If you hoist a thunk out of a context with `try - with -` in it, then you have to close it w.r.t. that. I can't think of a safe (by typing) and not heavy (monad heavy) solution. Maybe `fix_exn:exn -> exn` was a bit too blunt, but quite expressive ;-)

File: vernac/opaques.ml
Comment (gares) [inline vernac/opaques.ml:52]: The eta expansion is not needed, my bad I suggested it.

**Comment by ppedrot** [comment]
@gares Futures are much less interesting now since we're basically only considering a fixed code path in vernac/declare. We could simply defunctionalize the beast and reify the possible interactions as a nice AST. The only reason we had higher-order functions was to hide the horrible process state from the kernel.

File: vernac/opaques.ml
Comment (ppedrot) [reply vernac/opaques.ml:52]: I like it though, it makes the code more robust to stuff like optional arguments.

**Comment by gares** [comment]
Oh, I see. But then opaques should be even higher, in the DM (the STM for now), not in vernac.

**Comment by ppedrot** [comment]
CI green, AFAIU this is ready to merge.

**Comment by gares** [comment]
@coqbot merge now and let's party!

**Pull request merged**
